pmem2: add part coalescing feature

kswiecicki opened the pull request:
Depends on https://github.com/pmem/pmdk/pull/5128.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/5130)
<!-- Reviewable:end -->

pbalcer commented on doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md:
```diff
@@ -0,0 +1,54 @@
+The **pmemset_set_contiguous_part_coalescing**() sets part coalescing feature flag in
+pmemset to the provided value.
+
+When part coalescing feature is turned on each part mapped, using **pmemset_part_map**(3) function,
```
missing comma? This paragraph is hard to read :)

`When part coalescing is enabled, the pmemset_part_map function will ... `
?

Generally, when writing documentation, try to use active voice and use the simplest possible word variant.

pbalcer commented on src/libpmemset/part.c:
```diff
@@ -159,35 +160,56 @@ pmemset_part_map_new(struct pmemset_part_map **part_map,
+		pmem2_reserv = prev_pmap->pmem2_reserv;
+		offset = pmem2_vm_reservation_get_size(pmem2_reserv);
+		ret = pmem2_vm_reservation_extend(pmem2_reserv, part_size);
+	} else
```
use brackets uniformly in a series of statements. `if {} else {}` or `if ... else ...;`, never a mix.

pbalcer commented on src/libpmemset/part.c:
```diff
@@ -159,35 +160,56 @@ pmemset_part_map_new(struct pmemset_part_map **part_map,
 
-	pmem2_config_set_vm_reservation(pmem2_cfg, pmem2_reserv, 0);
+	/* coalesce part mappings into single vm reservation */
+	if (part_coalescing && prev_pmap) {
```
Initially, I envisioned that this option would only affect the failure behavior of this function - meaning that we always coalesce optionally (to increase performance), but if this is forced - we fail if coalescing is not possible.
Now I'm thinking all three behaviors make sense: never attempting to coalesce, attempting to coalesce, and forcing coalescing.

pbalcer commented on src/libpmemset/part.c:
```diff
@@ -159,35 +160,56 @@ pmemset_part_map_new(struct pmemset_part_map **part_map,
 		goto err_cfg_delete;
 	}
 
+	if (!(part_coalescing && prev_pmap)) {
```
create a variable like this: `int coalesce = part_coalescing && prev_pmap;` and then use it in the two conditions you have.

pbalcer commented on src/libpmemset/part.c:
```diff
@@ -202,9 +224,31 @@ void
+	size_t rsv_size = pmem2_vm_reservation_get_size(pmem2_reserv);
+
+	struct pmem2_map *map;
+	/* find mapping lying at the end of the reservation */
```
located

pbalcer commented on src/libpmemset/part.c:
```diff
@@ -202,9 +224,31 @@ void
+		pmem2_vm_reservation_delete(&pmem2_reserv);
+		Free(pmap);
+	} else {
+		/* mapping covers part of the vm reservation */
```
when does this happen? isn't the point of coalescing to have only one pmemset_part_map per reservation?

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 40 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r1-40:-MRxw2R9Dug7QQKMfngF:b-yvrg32) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L40)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

missing comma? This paragraph is hard to read :)

`When part coalescing is enabled, the pmemset_part_map function will ... `
?

Generally, when writing documentation, try to use active voice and use the simplest possible word variant.
</blockquote></details>

Done. Thanks for the advice.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 171 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-171:-MRxwG_rC3ox_y-XNQLz:b7b9pvh) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L171)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Initially, I envisioned that this option would only affect the failure behavior of this function - meaning that we always coalesce optionally (to increase performance), but if this is forced - we fail if coalescing is not possible.
Now I'm thinking all three behaviors make sense: never attempting to coalesce, attempting to coalesce, and forcing coalescing.
</blockquote></details>

Currently it's implemented to force coalescing when it's turned on. Is it viable behavior in regard to the overall libpmemset functionality?


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 175 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-175:-MRxwp7_4YcXtkdeipUZ:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L175)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

use brackets uniformly in a series of statements. `if {} else {}` or `if ... else ...;`, never a mix.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 203 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-203:-MRxwq5L6NVtPwZlnozE:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L203)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

create a variable like this: `int coalesce = part_coalescing && prev_pmap;` and then use it in the two conditions you have.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 233 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-233:-MRxwrFfEAelfNG7z1Y_:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L233)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

located
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/libpmemset/part.c, line 247 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-247:-MRxwsD83B-gvRAGcZwx:b-ae84vx) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L247)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

when does this happen? isn't the point of coalescing to have only one pmemset_part_map per reservation?
</blockquote></details>

There is only one part_map per vm reservation, but vm reservation can contain multiple pmem2_maps.
The corresponding *new* function creates a new part_map (therefore the vm reservation too) when coalescing is disabled. When it's enabled it modifies a previously mapped part_map  (if there is any, otherwise it works as without coalescing) by extending its vm reservation and appends a new part map to it.
So I made the *delete* function alike, it deletes whole vm reservation if it contains only pmem2 mapping that covers the whole reservation and only shrink the vm reservation when it contains multiple coalesced part mappings.
Both of those functions are internal and are used in pmemset.c file in *pmemset_part_map* function.


<!-- Sent from Reviewable.io -->

wlemkows commented on doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md:
```diff
@@ -0,0 +1,54 @@
+# DESCRIPTION #
+
+The **pmemset_set_contiguous_part_coalescing**() sets part coalescing feature flag in
+pmemset to the provided value.
```
to the provided value? you mean \*value\*?

wlemkows commented on doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md:
```diff
@@ -0,0 +1,54 @@
+pmemset to the provided value.
+
+When part coalescing is enabled, the **pmemset_part_map**(3) function will try to coalesce each
+new mapped part with the previously mapped part. Coalesced parts appear as single *struct pmemset_part_map*
```
What if it s not possible?

wlemkows commented on doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md:
```diff
@@ -0,0 +1,54 @@
+
+# RETURN VALUE
+
+The **pmemset_set_contiguous_part_coalescing**() function always returns 0.
```
config_set ?

wlemkows commented on src/test/pmemset_part/pmemset_part.c:
```diff
@@ -1,5 +1,5 @@
+/* Copyright 2020-2021, Intel Corporation */
 
 /*
  * pmemset_part.c -- pmemset_part unittests
```
do you have tests for this feature?

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 48 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-48:-MSb3klE80WYxiDvWIkc:baa1ri9) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L48)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

why not config_set ?
</blockquote></details>

We have concluded with Piotr that user can turn coalescing feature on/off between mappings.
Since *config_set* resides in the *pmemset* I would need to create internal 2 functions, one for the changing coalescing flag and one for retrieving the flag value in the *config_set*. Next i would need to create one public function for changing the flag value by the user, because the user couldn't use the config he created to change the flag value (the config is copied to the *pmemset* on the set creation).
It was easier to add this flag to the *pmemset* and since only *pmemset* needs to know the value of the flag, I can access it directly, without intermediate functions.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 38 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-38:-MSb6LJmBu_KPKJLZnYE:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L38)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

to the provided value? you mean \*value\*?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSb6iI_44xOw_f828fD:b8q9b99) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

What if it s not possible? 
</blockquote></details>

Setting the coalescing value is always possible. I'll write about coalescing feature in the *pmemset_part_map.3.md*.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/test/pmemset_part/pmemset_part.c, line 5 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDt-r3-5:-MSb6taW2YklHkHn3DeG:bopo2xb) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/src/test/pmemset_part/pmemset_part.c#L5)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

do you have tests for this feature?
</blockquote></details>

I've added the tests for the *vm_reservation_extend/shrink* and since they should be very similar to this I forgot. Thanks. I'll do it now.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130)* status: 0 of 31 files reviewed, 10 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)

___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSbHla_2b-HfxhYoo9Q:b-5ty916) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Setting the coalescing value is always possible. I'll write about coalescing feature in the *pmemset_part_map.3.md*.
</blockquote></details>

I meant "will try to coalesce each new mapped part with the previously mapped part.*"


<!-- Sent from Reviewable.io -->

wlemkows reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130)* status: 0 of 31 files reviewed, 9 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)

___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 38 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-38:-MSbJaRKBoxqQtp7BssV:b-6jrqdt) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L38)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

not done? I cannot see any changes here.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSbtOvoAjgvwQ6rQZvb:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

I meant "will try to coalesce each new mapped part with the previously mapped part.*"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 38 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-38:-MSbtQ7D0sgJh_DQcopp:b-j28f7j) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L38)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

not done? I cannot see any changes here.
</blockquote></details>

I was still working on the fixes and absentmindedly clicked done on this one.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/test/pmemset_part/pmemset_part.c, line 5 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDt-r3-5:-MSbteSiBsFnIWMdTH94:bqmjwcz) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/src/test/pmemset_part/pmemset_part.c#L5)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

I've added the tests for the *vm_reservation_extend/shrink* and since they should be very similar to this I forgot. Thanks. I'll do it now.
</blockquote></details>

Added.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130)* status: 0 of 34 files reviewed, 8 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)

___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSbuQWZ06NUwDmUDOx0:bsd755t) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks :) but my question is : "it **will try** to map the part directly", but there is no guarantee?  and user should now what happened if this "try" will fail or it cannot fail.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130)* status: 0 of 34 files reviewed, 9 unresolved discussions (waiting on @kswiecicki, @pbalcer, and @wlemkows)

___
*[src/test/pmemset_part/pmemset_part.c, line 1033 at r4](https://reviewable.io/reviews/pmem/pmdk/5130#-MSbvoyr4FstYIdL2TDv:-MSbvoyr4FstYIdL2TDw:b-m56hjz) ([raw file](https://github.com/pmem/pmdk/blob/b9f666028e23be4fae70f58303c7da87c7cb34bc/src/test/pmemset_part/pmemset_part.c#L1033)):*
> ```
> PMEMSET_E_CANNOT_COALESCE_PART
> ```
is this documented?


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSc-f42BASEJ07FUAsf:b-2o2zty) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Thanks :) but my question is : "it **will try** to map the part directly", but there is no guarantee?  and user should now what happened if this "try" will fail or it cannot fail.
</blockquote></details>

Okay, now I understand what you meant. I've mentioned it in the *set_config_contiguous_part_coalescing* manpage and elaborated on that in the *part_map* manpage.


<!-- Sent from Reviewable.io -->

kswiecicki commented:
___
*[src/test/pmemset_part/pmemset_part.c, line 1033 at r4](https://reviewable.io/reviews/pmem/pmdk/5130#-MSbvoyr4FstYIdL2TDv:-MSc0I7E4epvmcGoKSPR:b-jyesxm) ([raw file](https://github.com/pmem/pmdk/blob/b9f666028e23be4fae70f58303c7da87c7cb34bc/src/test/pmemset_part/pmemset_part.c#L1033)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

> ```
> PMEMSET_E_CANNOT_COALESCE_PART
> ```
is this documented?
</blockquote></details>

It's mentioned in the *part_map* manpage.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (commented):
> Reviewed 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130)* status: 1 of 34 files reviewed, 7 unresolved discussions (waiting on @kswiecicki and @pbalcer)

___
*[doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md, line 41 at r3](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDw-r3-41:-MSc0IOREG-MpxXUf0ks:b-ucfknh) ([raw file](https://github.com/pmem/pmdk/blob/18bf056befd25b3976caf6cc14c85c8e0d52d71b/doc/libpmemset/pmemset_set_contiguous_part_coalescing.3.md#L41)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Okay, now I understand what you meant. I've mentioned it in the *set_config_contiguous_part_coalescing* manpage and elaborated on that in the *part_map* manpage.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

pbalcer reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 31 files at r1, 8 of 15 files at r3, 10 of 18 files at r4, 1 of 3 files at r5, 7 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130#-:-MSm9wEx3FjUQRiOhskt:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @kswiecicki)

___
*[src/libpmemset/part.c, line 171 at r1](https://reviewable.io/reviews/pmem/pmdk/5130#-MRdKFkLnKRFcyIj6gDp-r1-171:-MSm8nZT2HMcRpx3UWte:b-o7i9x6) ([raw file](https://github.com/pmem/pmdk/blob/6a98e805463e131afa1bb0968e5ba55970ac799d/src/libpmemset/part.c#L171)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

Currently it's implemented to force coalescing when it's turned on. Is it viable behavior in regard to the overall libpmemset functionality?
</blockquote></details>

yes.
as a next step, we can think about changing this into an enum.


<!-- Sent from Reviewable.io -->

wlemkows reviewed (commented):
With coalescing feature I am not able to map multiple parts.
I got PMEMSET_E_CANNOT_COALESCE_PART_MAPS.
In fact it fails on pmem2_vm_reservation_extend->vm_reservation_reserve_memory with MAP_FAILED
and errno == EEXIST.

> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130#-:-MT0q9ia09NtjFN8GMP2:bir7sl2)* status: all files reviewed, 2 unresolved discussions (waiting on @kswiecicki)



<!-- Sent from Reviewable.io -->

pbalcer reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/pmdk/5130#-:-MUDFUmh6z-J4RvLTLfy:bnfp4nl)* status: 16 of 18 files reviewed, 1 unresolved discussion (waiting on @kswiecicki and @pbalcer)



<!-- Sent from Reviewable.io -->
